text/3550-new-range.md

```diff
@@ -0,0 +1,441 @@
+
+These types implement the `IntoIterator` trait, enabling their use directly in a `for` loop:
+```rust
+for n in 0..5 {
+    // `n` = 0, 1, 2, 3, 4
+}
+```
+
+All range types are `Copy` when the bounds are `Copy`, allowing easy reuse:
+```rust
+let range = 0..5;
+
+if a_slice[range].contains(x) {
+    // ...
+}
+if b_slice[range].contains(y) {
+    // ...
+}
+```
+
+For convenience, several commonly-used methods from `Iterator` are present as inherent functions on the range types:
+```rust
+for n in (1..).map(|x| x * 2) {
+    // n = 2, 4, 6, 8, 10, ...
+}
+for n in (0..5).rev() {
+    // n = 4, 3, 2, 1, 0
+}
+```
+
+## Legacy Range Types
+
+In Rust editions prior to 2024, `a..b`, `a..=b`, and `a..` resolved to a different set of types (now found in `std::ops::range::legacy`). These legacy range types did not implement `Copy`, and implemented `Iterator` directly (rather than `IntoIterator`).
+
+This meant that any `Iterator` method could be called on those range types:
+```rust
+let mut range = 0..5;
+assert_eq!(range.next(), Some(0));
+range.for_each(|n| {
+    // n = 1, 2, 3, 4
+});
+```
+
+There exist `From` impls for converting from the new range types to the legacy range types.
+
+### Migrating
+
+`cargo fix --edition 2024` can handle migrating many use cases to the new range types. This includes most places where a `RangeBounds` or `Iterator` is expected:
+```rust
+pub fn takes_range(range: impl std::ops::RangeBounds<usize>) { ... }
+takes_range(0..5); // No changes necessary
+```
+
+And most places where `Iterator` methods were used directly on the range:
+```rust
+for n in (0..5).rev() { ... } // No changes necessary
+
+// Before
+(0..5).for_each(...);
+// After
+(0..5).into_iter().for_each(...); // Add `.into_iter()`
+```
+
+But it is impossible to generally handle everything, and there are cases which fall back to converting to the legacy types:
+
+```rust
+// Before
+pub fn takes_range(range: std::ops::Range<usize>) { ... }
+takes_range(0..5);
+// After
+pub fn takes_range(range: std::ops::range::legacy::Range<usize>) { ... }
+takes_range(std::ops::range::legacy::Range::from(0..5));
+```
+
+To reduce the need for conversions, we recommend the following:
+
+- Change any function parameters from legacy `Range*` types to `impl RangeBounds<_>`  
+  Or where `RangeBounds` is not applicable, `impl Into<Range*>`
+
+```rust
+// Before
+pub fn takes_range(range: std::ops::Range<usize>) { ... }
+
+// After
+pub fn takes_range(range: impl std::ops::RangeBounds<usize>) { ... }
+// Or
+pub fn takes_range(range: impl Into<std::ops::range::legacy::Range<usize>>) { ... }
+```
+
+- Change any trait bounds that assume `Range*: Iterator` to use `IntoIterator` instead  
+  This is fully backwards-compatible, thanks to the [blanket `impl<I: Iterator> IntoIterator for I`](https://doc.rust-lang.org/stable/std/iter/trait.IntoIterator.html#impl-IntoIterator-for-I)
+
```

Comment:
This section is just overall confusing to me.

One big point is that `cargo fix` never guarantees that things will always work when converting between editions. We try to cover everything, but ultimately, people will have to manually fix some things when moving over, especially if there's ambiguity over what a developer meant by code.

This seems to fall into the same category. I personally would say that anyone specifically accepting `Range<usize>` instead of `impl RangeBounds<usize>` has a reason to do so, and simply telling them to switch to using `impl RangeBounds<usize>` is probably bad advice. I do think that recommending to users to use `impl RangeBounds<usize>` whenever they can is good advice, just not that they should mindlessly convert everything to use that without considering the implications.